docs/docs/dev_docs/contracts/resources/common_patterns/main.md

```diff
@@ -0,0 +1,139 @@
+---
+title: Common Patterns and Anti-Patterns
+---
+
+There are many common patterns have been devised by the Aztec core engineering team and the work of the external community as we build Aztec.nr contracts internally (see some of them [here](https://github.com/AztecProtocol/aztec-packages/tree/master/yarn-project/noir-contracts)).
+
+This doc aims to summarize some of them!
+
+Similarly we have discovered some anti-patterns too (like privacy leakage) that we will point out here!
+
+## Common Patterns
+### Safe Math and SafeU120
+Field operations may overflow/underflow. Hence we have built a SafeMath library that you can use [based on instructions here](../dependencies.md#safe-math)
+
+Comparison on Field is also not possible today. For such cases, we recommend using u120, which is wrapped under the SafeU120 class found in the SafeMath library.
+
+### Approving another user/contract to call a function on your behalf
+We call this the "authentication witness" pattern or authwit for short.
+- Approve someone in private domain:
+#include_code authwit_to_another_sc /yarn-project/end-to-end/src/e2e_cross_chain_messaging.test.ts typescript
+
+Here you approve a contract to burn funds on your behalf.
+
+- Approve in public domain:
+#include_code authwit_public_transfer_example /yarn-project/end-to-end/src/e2e_token_contract.test.ts typescript
+
+Here you approve someone to transfer funds publicly on your behalf
+
+### Prevent the same user flow from happening twice or avoiding replay attacks using nullifiers 
+E.g. you don't want a user to subscribe once they have subscribed already. Or you don't want them to vote twice once they have done that. How do you prevent this?
+
+Emit a nullifier to prevent them from such "replay attacks". This is also why in authwit, we emit a nullifier, to prevent someone from reusing their approval. 
+
+#include_code assert_valid_authwit_public /yarn-project/aztec-nr/authwit/src/auth.nr rust
+
+Note be careful to ensure that the nullifier is not deterministic and that no one could do a preimage analysis attack. More in [the anti pattern section on deterministic nullifiers](#deterministic-nullifiers)
+
+Note - you could also create a note and send it to the user. The problem is there is nothing stopping the user from not presenting this note when they next interact with the function. 
+
+### Reading public storage in private
+You can't read public storage in private domain. But nevertheless reading public storage is desirable. There are two ways:
+
+1. You pass the data as a parameter to your private method and later assert in public that the data is correct. E.g.:
```

Comment:
Should we warn that this leaks information about the private function that was executed (and the data which has been read)? This almost seems like an anti-pattern for privacy?